packages/app/src/sandbox/compile.ts

```diff
@@ -890,11 +890,9 @@ async function compile(opts: CompileOptions) {
 
     setTimeout(async () => {
       try {
-        testRunner =
-          testRunner ||
-          (await import('./eval/tests/jest-lite')
-            .then(s => s.default)
-            .then(TestRunner => new TestRunner(manager)));
+        const jestLiteModule = await import('./eval/tests/jest-lite');
```

Comment:
Maybe I'm misreading this. But isn't it doing exactly the same thing but in a different order? 

Plus, maybe we could make import conditional in order to prevent loading/fetching it once again:

```js
if(!testRunner) {
	const jestLiteModule = await import('./eval/tests/jest-lite');
	const TestRunner = jestLiteModule.default;
	testRunner = new TestRunner(manager);
}
```

What do you think?

Reply:
Hey @danilowoz, so it's subtle, but they're not exactly doing the same thing. The difference comes down to _when_ the `testRunner` variable conditional is checked.

In the current code and the code that you suggested, we're checking whether `testRunner` exists _before_ we call `await import('./eval/tests/jest-lite');`. What's happening is that when the first `compile` call reaches the `await import('./eval/tests/jest-lite');` line, the process suspends its execution. Before the import is returned and `testRunner` can be set, the second `compile` call reaches the _same_ `await import('./eval/tests/jest-lite');` line.

Because the first call was suspended and `testRunner` hasn't been set yet, the second call will _also_ fail this check and start the import process again. This results in `new TestRunner(manager)` being called twice, leading to the bug we're seeing.

The difference with the code in the PR is that we check if `testRunner` has been set _after_ the `await import('./eval/tests/jest-lite');` line (i.e. `testRunner = testRunner || new TestRunner(manager);`). Here's a rough explanation of what I think is happening in this scenario:

1. First `compile` call reaches the `await import('./eval/tests/jest-lite');` line
2. First `compile` call suspends its execution because of the `await`
3. Second `compile` call reaches the `await import('./eval/tests/jest-lite');` line
4. Second `compile` call suspends its execution because of the `await`
5. First `compile` call resumes execution and  sets `testRunner` because it's `undefined` and then it finishes executing
6. Second `compile` call resumes execution and **doesn't** set `testRunner` because we do the check at this point _after_ Step 5 finishes

In all 3 scenarios, we're still calling the `await import('./eval/tests/jest-lite');` twice, but by moving the check to see if `testRunner` exists to _after_ the `await`, we're able to skip the second call to `new TestRunner(manager);`.

I put a minimal example together that outlines each of the 3 scenarios: https://codesandbox.io/s/mystifying-dream-lcfdtz?file=/src/index.js. If you scroll to the bottom of the sandbox, there are 3 versions of the main function that represent the 3 different scenarios we're talking about:
1. The current code
2. The code you're suggesting
3. The code in the PR

If you open the console and try uncommenting each of those main functions, you should be able to see how the code flow is logged. The key part is that only the last example calls the `createTestRunner` function once.

One other thing to note is that even though we're calling `await import("./createTestRunner")` twice in the example CodeSandbox, the `console.log` at the top level of the imported file is only executed once. I can't say for sure, but it looks like even though we `import` the same module more than once, it doesn't actually load it multiple times (maybe it's served from a cache).

Reply:
Gotcha!! Great example, btw.

Let's get this merged